docs/src/pages/components/data-grid/components/components.md

```diff
@@ -7,10 +7,83 @@ components: DataGrid, XGrid
 
 <p class="description">The grid is highly customizable. Override components using the <code>components</code> prop.</p>
 
-As part of the customization API, the grid allows you to replace and override nested components with the `components` prop.
-The prop accepts an object of type [GridSlotsComponent](/api/data-grid/#slots) .
+## Overriding components
 
-## Toolbar
+As part of the customization API, the grid allows you to override internal components with the `components` prop.
+The prop accepts an object of type [`GridSlotsComponent`](/api/data-grid/#slots).
+
+If you wish to pass additional props in a component slot, you can do it using the `componentsProps` prop. This prop is of type `GridSlotsComponentsProps`.
+
+As an example, you could override the column menu and pass additional props as below.
+
+```jsx
+<DataGrid
+  rows={rows}
+  columns={columns}
+  components={{
+    ColumnMenu: MyCustomColumnMenu,
+  }}
+  componentsProps={{
+    columnMenu: { background: 'red', counter: rows.length },
+  }}
+/>
+```
+
+**Note** The casing is different between the `components` (ColumnMenu) and `componentsProps` (columnMenu) props.
+
+### Getting props
+
+While overriding component slots, you might need to access the grid data.
+Therefore, the grid expose a `useGridSlotComponentProps` hook which allows to retrieve the following props.
```

Comment:
There's still the void space pointed in https://github.com/mui-org/material-ui-x/pull/1382#discussion_r631246011.

```suggestion
Therefore, the grid exposes a `useGridSlotComponentProps` hook which allows to retrieve the following props.
```